### PR TITLE
[mlir] expose transform dialect symbol merge to python

### PR DESCRIPTION
**Pull request by ftynse**
This functionality is available in C++, make it available in Python directly to operate on transform modules.

### CODE REVIEW

**Review by stellaraccident — COMMENTED** [review]
Drive-by: why the copy vs move of operations?

For reference, one of several of these I've written in Python over the years.

**Comment by ftynse** [comment]
I don't see how one can implement a move in a reasonable way in Python. It is possible to mark the moved-from operation as invalid at the bindings level and raise an error on the second access, but it isn't something the language naturally has and is therefore surprising. OTOH, the anticipated use case is "load library A, merge symbols from A into script X, merge symbols from B into script Y, ..." so we don't want to consume the library.

File: mlir/lib/Bindings/Python/TransformInterpreter.cpp [hunk 1]
```diff
@@ -82,6 +82,21 @@ static void populateTransformInterpreterSubmodule(py::module &m) {
...
+      [](MlirOperation target, MlirOperation other) {
+        mlir::python::CollectDiagnosticsToStringScope scope(
+            mlirOperationGetContext(target));
+
+        MlirLogicalResult result = mlirMergeSymbolsIntoFromClone(target, other);
+        if (mlirLogicalResultIsSuccess(result))
+          return;
+
+        throw py::value_error("Failed to merge symbols.\nDiagnostic message " +
+                              scope.takeMessage());
```
Comment (makslevental) [inline mlir/lib/Bindings/Python/TransformInterpreter.cpp:97]: ultra nit: don't we usually spell this as "if(failed) return error"?

**Review by makslevental — APPROVED** [review]
LGTM

**Comment by ftynse** [comment]
> A nit could be "well then it should be named copy_symbols_into" but then you lose the clear/loud indication that things will be renamed.

Renamed to `copy_symbols_and_merge_into`.

**Pull request merged**